File: app/src/main/java/com/faendir/lightning_launcher/multitool/drawer/RestorePreference.java
package com.faendir.lightning_launcher.multitool.drawer;

import android.content.ComponentName;
import android.content.Context;
import android.content.pm.PackageManager;
import android.util.AttributeSet;
import androidx.preference.MultiSelectListPreference;
import java9.util.stream.StreamSupport;

import java.util.Collection;
import java.util.Set;

import static com.faendir.lightning_launcher.multitool.util.LambdaUtils.exceptionToOptional;

/**
 * @author F43nd1r
 * @since 13.11.2016
 */

public class RestorePreference extends MultiSelectListPreference {
    private final PackageManager pm;

    public RestorePreference(Context context, AttributeSet attrs) {
        super(context, attrs);
        setPersistent(true);
        pm = getContext().getPackageManager();
    }

    private String getLabelForComponent(String flatComponent) {
        return exceptionToOptional(pm::getActivityInfo).apply(ComponentName.unflattenFromString(flatComponent), 0).map(info -> info.loadLabel(pm).toString()).orElse("");
    }

    @Override
    public boolean callChangeListener(Object newValue) {
        Set<String> values = getValues();
        //noinspection SuspiciousMethodCalls
        values.removeAll((Collection<?>) newValue);
        super.callChangeListener(values);
        setValues(values);
        return false;
    }

    @Override
    protected void onSetInitialValue(Object defaultValue) {
        //noinspection unchecked
        Set<String> values = getPersistedStringSet((Set<String>) defaultValue);
        setEntryValues(values.toArray(new String[0]));
        setEntries(StreamSupport.stream(values).map(this::getLabelForComponent).toArray(String[]::new));
        setValues(values);
    }
}

File: app/src/main/java/com/faendir/lightning_launcher/multitool/drawer/RestorePreference.kt
package com.faendir.lightning_launcher.multitool.drawer

import android.content.ComponentName
import android.content.Context
import android.content.pm.PackageManager
import android.util.AttributeSet
import androidx.preference.MultiSelectListPreference

/**
 * @author F43nd1r
 * @since 13.11.2016
 */

class RestorePreference(context: Context, attrs: AttributeSet) : MultiSelectListPreference(context, attrs) {
    private val pm: PackageManager = getContext().packageManager

    init {
        isPersistent = true
        entries = arrayOfNulls(0)
        entryValues = arrayOfNulls(0)
    }

    private fun getLabelForComponent(flatComponent: String): String = pm.getActivityInfo(ComponentName.unflattenFromString(flatComponent), 0)?.loadLabel(pm).toString()

    override fun callChangeListener(newValue: Any): Boolean {
        val values = values
        values.removeAll(newValue as Collection<*>)
        super.callChangeListener(values)
        setValues(values)
        return false
    }

    override fun onSetInitialValue(defaultValue: Any?) {
        val values: Set<String>? = if (defaultValue is Collection<*>) getPersistedStringSet(defaultValue.map { it.toString() }.toMutableSet()) else emptySet()
        entryValues = values?.toTypedArray()
        entries = values?.map { this.getLabelForComponent(it) }?.toTypedArray()
        setValues(values)
    }
}
